rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+    pub block_number: BlockNumber,
+    /// Timestamp of the beginning of the last block creation attempt.
+    pub block_timestamp: BlockTimestamp,
+    /// The sequencer address of this block.
+    pub sequencer_address: ContractAddress,
+}
+
+type RpcResponse = ureq::Response;
+
+impl RpcState {
+    pub fn new(chain: RpcChain, block: BlockValue) -> Self {
+        if env::var("INFURA_API_KEY").is_err() {
+            dotenv().expect("Missing .env file");
+        }
+        Self {
+            chain,
+            api_key: env::var("INFURA_API_KEY")
+                .expect("Missing API Key in environment: INFURA_API_KEY"),
+            block,
+        }
+    }
+
+    fn rpc_call<T: for<'a> Deserialize<'a>>(
+        &self,
+        params: &serde_json::Value,
+    ) -> Result<T, RpcError> {
+        Self::deserialize_call(self.rpc_call_no_deserialize(params)?)
+    }
+
+    fn rpc_call_no_deserialize(&self, params: &serde_json::Value) -> Result<RpcResponse, RpcError> {
+        ureq::post(&format!(
+            "https://{}.infura.io/v3/{}",
+            self.chain, self.api_key
+        ))
+        .set("Content-Type", "application/json")
+        .set("accept", "application/json")
+        .send_json(params)
+        .map_err(|err| RpcError::Request(err.to_string()))
+    }
+
+    fn deserialize_call<T: for<'a> Deserialize<'a>>(response: RpcResponse) -> Result<T, RpcError> {
+        let response = response.into_string().map_err(|err| {
+            RpcError::Cast("Response".to_owned(), "String".to_owned(), err.to_string())
+        })?;
+        serde_json::from_str(&response).map_err(|err| RpcError::RpcCall(err.to_string()))
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct RpcResult<T> {
+    result: T,
+}
+
+#[derive(Debug)]
+pub struct TransactionTrace {
+    pub validate_invocation: RpcCallInfo,
+    pub function_invocation: RpcCallInfo,
+    pub fee_transfer_invocation: RpcCallInfo,
+    pub signature: Vec<StarkFelt>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq)]
+pub struct RpcExecutionResources {
+    pub n_steps: usize,
+    pub n_memory_holes: usize,
+    pub builtin_instance_counter: HashMap<String, usize>,
+}
+
+#[derive(Debug)]
+pub struct RpcCallInfo {
+    pub execution_resources: VmExecutionResources,
+    pub retdata: Option<Vec<StarkFelt>>,
+    pub calldata: Option<Vec<StarkFelt>>,
+    pub internal_calls: Vec<RpcCallInfo>,
+}
+
+#[allow(unused)]
+#[derive(Debug, Deserialize)]
+pub struct RpcTransactionReceipt {
+    #[serde(deserialize_with = "actual_fee_deser")]
+    actual_fee: u128,
+    block_hash: StarkHash,
+    block_number: u64,
+    execution_status: String,
+    #[serde(rename = "type")]
+    tx_type: String,
+}
+
+fn actual_fee_deser<'de, D>(deserializer: D) -> Result<u128, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let hex: &str = Deserialize::deserialize(deserializer)?;
+    Ok(u128::from_str_radix(&hex[2..], 16).unwrap())
```

Comment:
Is not checking the prefix intentional? (the `0x` prefix)

Reply:
I didn't deem it necessary to add a check because every `StarkFelt` is deserialized assuming the prefix as by implementation in `starknet_api` (see `PrefixedBytesAsHex`), and in this case because the actual fee won't ever be greater than the field's prime, it's safe to truncate, which is equivalent to directly interpreting as a `u128`.